### DISCUSSION TITLE
Do you can concurently launch several chrome instances each with own user-data-dir

### DISCUSSION THREAD

**Discussion by MikhailKlemin** [go-rod/rod#429] [Q&A] [answered]
What I want is to have several instances running each with it's own datadir: 
Say create concurrently several of this with different profiles? 

I guess I cannot? Because new instance will mess up with already launched? 
 
```go
func create(profile string) *rod.Browser {
	l := launcher.NewUserMode().UserDataDir(profile).
		Headless(false).
		Set("blink-settings", "imagesEnabled=false").
		Devtools(false)

	url := l.MustLaunch()
	//l.Cleanup()

	browser := rod.New().
		ControlURL(url).
		Trace(false).
		SlowMotion(1 * time.Second).
		MustConnect()

	return browser

}

```

**Comment by ysmood** [comment]
Yes, you can. Check the unit test of Rod itself, it will run each test concurrently with multiple browsers.

https://github.com/go-rod/rod/blob/0998fccf111871a7e2168a4e717bd522f31e75b6/setup_test.go#L60

**Reply by MikhailKlemin** [reply]
If I just use standard launcher (I mean without `NewUserMode` -- it's all works fine):

```go
// this will work fine:
func create(profile string) *rod.Browser {
	//l := launcher.NewUserMode().UserDataDir(profile).
	path, _ := launcher.LookPath()
	l := launcher.New().Bin(path).
		Headless(false).
		Set("blink-settings", "imagesEnabled=false").
		Devtools(false)

	url := l.MustLaunch()
	browser := rod.New().
		ControlURL(url).
		Trace(false).
		SlowMotion(1 * time.Second).
		MustConnect()

	return browser
}
```

But code bellow with parallel produces unpredictable (at my glance) behavior:

```go 
func create(profile string) *rod.Browser {
	l := launcher.NewUserMode().UserDataDir(profile).
		Headless(false).
		Set("blink-settings", "imagesEnabled=false").
		Devtools(false)

	url := l.MustLaunch()

	browser := rod.New().
		ControlURL(url).
		Trace(false).
		SlowMotion(1 * time.Second).
		MustConnect()

	return browser

}

func main() {
	sem := make(chan bool, 5)
	for i := 0; i < 5; i++ {
		sem <- true
		go func(i int) {
			defer func() { <-sem }()
			b := create(fmt.Sprintf("/tmp/%d", i))
			_, err := b.Page(proto.TargetCreateTarget{})
			if err != nil {
				log.Fatal(err)
			}
			time.Sleep(5 * time.Second)
			fmt.Printf("Closing browser # %d\n", i)
			err = b.Close()
			if err != nil {
				log.Fatal(err)
			}

		}(i)
	}

	for i := 0; i < cap(sem); i++ {
		sem <- true
	}

}
```
Some of issues might be caused, by pop window Chrome launches asking to make it Default Browser.

**Reply by ysmood** [reply]
Yes, using user-mode is crazy, I will never do that, chromium is not designed to do that, unexpected behavior is not surprising for me.

**Reply by MikhailKlemin** [reply]
Maybe I do not understand something obvious, I only need to have several instances running in parallel with separate UserDataDirs.

**Reply by ysmood** [reply]
Why not use `launcher.New()`?

**Reply by MikhailKlemin** [reply]
I want to check same website in parallel with different sets of data stored in cookies, web storage and other persistent options if they are exists. 
I thought that getting each chrome instance with it's own profile would solve that.

**Comment by ysmood** [comment, marked as answer]
```go
func create(profile string) *rod.Browser {
	u := launcher.New().UserDataDir(profile).
		Set("blink-settings", "imagesEnabled=false").
		MustLaunch()

	return rod.New().ControlURL(u).MustConnect()
}

func main() {
	wg := sync.WaitGroup{}
	for range "....." {
		wg.Add(1)
		go func() {
			defer wg.Done()

			p := filepath.Join(os.TempDir(), utils.RandString(8))

			b := create(p)
			b.MustPage()
			fmt.Println("Browser started:", p)
			time.Sleep(5 * time.Second)
			b.MustClose()
			fmt.Println("Browser stopped:", p)
		}()
	}
	wg.Wait()
}

```

**Reply by MikhailKlemin** [reply]
This works fine, thanks a lot! 
I have tried almost the same but was getting panic error `panic: [launcher] Failed to get the debug url: Opening in existing browser session.`   and then I started looking I guess into wrong direction!
